lib/appium.js

```diff
@@ -344,35 +346,142 @@ class AppiumDriver extends BaseDriver {
     }
   }
 
+  pluginsToHandleCmd (cmd) {
+    return this.plugins.filter((p) => p.commands === true || _.includes(p.commands, cmd));
+  }
+
   async executeCommand (cmd, ...args) {
-    // getStatus command should not be put into queue. If we do it as part of super.executeCommand, it will be added to queue.
-    // There will be lot of status commands in queue during createSession command, as createSession can take up to or more than a minute.
-    if (cmd === 'getStatus') {
-      return await this.getStatus();
+    // We have basically three cases for how to handle commands:
+    // 1. handle getStatus (we do this as a special out of band case so it doesn't get added to an
+    //    execution queue, and can be called while e.g. createSession is in progress)
+    // 2. handle commands that this umbrella driver should handle, rather than the actual session
+    //    driver (for example, deleteSession, or other non-session commands)
+    // 3. handle session driver commands.
+    // The tricky part is that because we support command plugins, we need to wrap any of these
+    // cases with plugin handling.
+
+    const isGetStatus = cmd === 'getStatus';
+    const isUmbrellaCmd = !isGetStatus && isAppiumDriverCommand(cmd);
+    const isSessionCmd = !isGetStatus && !isUmbrellaCmd;
+
+    // first do some error checking. If we're requesting a session command execution, then make
+    // sure that session actually exists on the session driver, and set the session driver itself
+    let sessionId = null;
+    let dstSession = null;
+    let protocol = null;
+    if (isSessionCmd) {
+      sessionId = _.last(args);
+      dstSession = await sessionsListGuard.acquire(AppiumDriver.name, () => this.sessions[sessionId]);
+      if (!dstSession) {
+        throw new Error(`The session with id '${sessionId}' does not exist`);
+      }
+      // now save the response protocol given that the session driver's protocol might differ
+      protocol = dstSession.protocol;
     }
 
-    if (isAppiumDriverCommand(cmd)) {
-      return await super.executeCommand(cmd, ...args);
-    }
+    // now we define a 'handled' object which will keep track of which plugins have handled this
+    // command. we care about this because (a) multiple plugins can handle the same command, and
+    // (b) there's no guarantee that a plugin will actually call the next() method which runs the
+    // original command execution. This results in a situation where the command might be handled
+    // by some but not all plugins, or by plugin(s) but not by the default behavior. So start out
+    // this object declaring that the default handler has not been executed.
+    const handled = {default: false};
```

Comment:
can we have a better name for this var, that makes the fact it is used by plugins more obvious?